### ISSUE TITLE
Async calls with continuations are not memoized correctly with a memo effect

### ISSUE THREAD

**Issue by a-shkarupin** [atnos-org/eff#48]
Consider this example for `AsyncFutureInterpreterSpec`:
```scala
def e14 = {
    var invocationsNumber = 0
    val cache = ConcurrentHashMapCache()

    type S = Fx.fx2[Memoized, Async]
    val intAsync: Eff[S, Int] = asyncFork[S, String]("a").flatMap( x => asyncFork[S, Int]({ invocationsNumber += 1; 1 }))
    def makeRequest = asyncMemoized("only once", intAsync)

    (makeRequest >> makeRequest).runAsyncMemo(cache).runAsyncFuture must be_==(1).await
    invocationsNumber must be_==(1)
  }
```
Looks like `flatMap`ping causes the result of the first async to be cached. I would expect `1` to be cached, however `"a"` is cached instead:
```
[error]  x Async calls with continuations can be memoized with a memo effect
[error]  'a' is not equal to '1' (AsyncFutureInterpreterSpec.scala:179)

```

**Comment by etorreborre** [comment]
The memoization is totally not-working at the moment. I have a fix which I will finish tomorrow.

**Comment by etorreborre** [comment]
It just took me 4 days to really "finish" the fix... It is now available in `2.3.0`.
